File: Source/Game/Management/Progress.cs
using System;
using System.Collections.Generic;
using System.Linq;
using WCSharp.Api;

public sealed class Progress : IProgressService
{
    private static Progress _instance;
    public static Progress Instance => _instance
        ?? throw new InvalidOperationException("Progress has not been initialized. Call Progress.Initialize() before accessing the instance.");

    public Dictionary<int, float> DistancesFromStart { get; private set; } = new Dictionary<int, float>();
    private timer TeamProgTimer { get; set; } = timer.Create();

    public static void Initialize()
    {
        _instance ??= new Progress();
    }

    private Progress()
    {
        CalculateTotalDistance();
        if (Gamemode.CurrentGameMode != GameMode.Team) return;
        TeamProgTimer.Start(0.2f, true, TeamProgressTracker);
    }

    public float CalculateProgress(ProgressRequest request)
    {
        return CalculatePlayerProgress(request);
    }

    private void TeamProgressTracker()
    {
        if (!Globals.GAME_ACTIVE) return;
        try
        {
            var allTeams = TeamRegistry.All;
            for (int i = 0; i < allTeams.Count; i++)
            {
                var team = allTeams[i];
                team.UpdateRoundProgress(Globals.ROUND, CalculateTeamProgress(team));
            }
            //TeamsMultiboard.UpdateTeamStatsMB();
        }
        catch (Exception e)
        {
            Logger.Warning($"Error in TeamProgressTracker. {e.Message}");
        }
    }

    private string CalculateTeamProgress(Team Team)
    {
        float totalProgress = 0.0f;

        if (Team.Teammembers.Count == 0) return "0.00";

        for (int i = 0; i < Team.Teammembers.Count; i++)
        {
            var kitty = Team.Teammembers[i];
            if (kitty.Finished) totalProgress += 100.00f;
            else totalProgress += kitty.TimeProg.GetRoundProgress(Globals.ROUND);

        }

        return (totalProgress / Team.Teammembers.Count).ToString("F2");
    }

    private float CalculatePlayerProgress(ProgressRequest request)
    {
        try
        {
            // Once a kitty has finished (entered the final safezone or the victory area), always report 100 progress.
            if (request.Finished || HasReachedFinalSafezone(request)) return 100.0f;

            var currentSafezone = request.ProgressZone;
            if (Globals.SAFE_ZONES[0].Rectangle.Contains(request.UnitX, request.UnitY)) return 0.0f; // if at start, 0 progress
            var currentProgress = DistanceBetweenPoints(request.UnitX, request.UnitY,
                ProgressPointHelper.Points[request.CurrentPoint].X, ProgressPointHelper.Points[request.CurrentPoint].Y);
            var totalProgress = DistancesFromStart[currentSafezone] + currentProgress;

            var progress = totalProgress / DistancesFromStart[RegionList.PathingPoints.Length - 1] * 100;
            if (progress > 100) progress = 100.00f;

            return progress;
        }
        catch (Exception e)
        {
            Logger.Warning($"Error in CalculatePlayerProgress. {e.Message}");
            return 0.0f;
        }
    }

    private static bool HasReachedFinalSafezone(ProgressRequest request)
    {
        return Globals.SAFE_ZONES[Globals.SAFE_ZONES.Count - 1].Rectangle.Contains(request.UnitX, request.UnitY)
            || Regions.Victory_Area.Contains(request.UnitX, request.UnitY);
    }

    public float CalculateNitroPacerProgress()
    {
        var nitroKitty = NitroPacer.Unit;
        var currentSafezone = NitroPacer.GetCurrentCheckpoint();
        if (Globals.SAFE_ZONES[0].Rectangle.Contains(nitroKitty.X, nitroKitty.Y)) return 0.0f; // if at start, 0 progress
        if (Globals.SAFE_ZONES[Globals.SAFE_ZONES.Count - 1].Rectangle.Contains(nitroKitty.X, nitroKitty.Y)) return 100.0f; // if at end.. 100 progress
        var currentProgress = DistanceBetweenPoints(nitroKitty.X, nitroKitty.Y,
            ProgressPointHelper.Points[currentSafezone].X, ProgressPointHelper.Points[currentSafezone].Y);
        var totalProgress = DistancesFromStart[currentSafezone] + currentProgress;

        return totalProgress;
    }

    private void CalculateTotalDistance()
    {
        try
        {
            if (RegionList.PathingPoints == null || RegionList.PathingPoints.Length == 0)
            {
                Logger.Warning("PathingPoints list is null or empty.");
                return;
            }

            var totalDistance = 0.0f;
            var count = 0;
            DistancesFromStart.Add(0, 0.0f);
            foreach (var pathPoint in RegionList.PathingPoints)
            {
                if (count >= RegionList.PathingPoints.Length - 1) break;
                var nextPathPoint = RegionList.PathingPoints[count + 1];
                totalDistance += DistanceBetweenPoints(pathPoint.Rect.CenterX, pathPoint.Rect.CenterY, nextPathPoint.Rect.CenterX, nextPathPoint.Rect.CenterY);
                if (!DistancesFromStart.ContainsKey(count + 1)) DistancesFromStart.Add(count + 1, totalDistance);
                count++;
            }
        }
        catch (Exception e)
        {
            Logger.Warning($"Error in CalculateTotalDistance. {e.Message}");
            throw;
        }
    }

    private static float DistanceBetweenPoints(float x1, float y1, float x2, float y2)
    {
        return Math.Abs(x1 - x2) > Math.Abs(y1 - y2) ? Math.Abs(x1 - x2) : Math.Abs(y1 - y2);
    }
}

File: src/Game/Management/Progress.ts
export class Progress {
    public static DistancesFromStart: Map<number, number> = new Map()
    private static TeamProgTimer = Timer.create()

    public static Initialize() {
        CalculateTotalDistance()
        if (Gamemode.CurrentGameMode != GameMode.TeamTournament) return
        TeamProgTimer.start(0.2, true, TeamProgressTracker)
    }

    public static CalculateProgress(kitty: Kitty) {
        let round = Globals.ROUND
        kitty.TimeProg.SetRoundProgress(round, CalculatePlayerProgress(kitty))
    }

    private static TeamProgressTracker() {
        if (!Globals.GAME_ACTIVE) return
        try {
            for (let i: number = 0; i < Globals.ALL_TEAMS_LIST.length; i++) {
                let team = Globals.ALL_TEAMS_LIST[i]
                team.UpdateRoundProgress(Globals.ROUND, CalculateTeamProgress(team))
            }
            TeamsMultiboard.UpdateTeamStatsMB()
        } catch (e: any) {
            Logger.Warning('Error in TeamProgressTracker. {e.Message}')
        }
    }

    private static CalculateTeamProgress(Team: Team) {
        let totalProgress: number = 0.0

        if (Team.Teammembers.length == 0) return '0.00'

        for (let i: number = 0; i < Team.Teammembers.length; i++) {
            let player = Team.Teammembers[i]
            totalProgress += Globals.ALL_KITTIES.get(player)!.TimeProg.GetRoundProgress(Globals.ROUND)
        }

        return (totalProgress / Team.Teammembers.length).ToString('F2')
    }

    private static CalculatePlayerProgress(kitty: Kitty) {
        try {
            let currentSafezone = kitty.ProgressZone
            if (Globals.SAFE_ZONES[Globals.SAFE_ZONES.length - 1].Rectangle.includes(kitty.Unit.x, kitty.Unit.y))
                return 100.0 // if at end.. 100 progress
            if (Regions.Victory_Area.includes(kitty.Unit.x, kitty.Unit.y)) return 100.0 // if in victory area, 100 progress
            if (Globals.SAFE_ZONES[0].Rectangle.includes(kitty.Unit.x, kitty.Unit.y) && !kitty.Finished) return 0.0 // if at start, 0 progress
            if (kitty.isAlive() && kitty.Finished) return 100.0
            let currentProgress = DistanceBetweenPoints(
                kitty.Unit.x,
                kitty.Unit.y,
                ProgressPointHelper.Points[kitty.ProgressHelper.CurrentPoint].x,
                ProgressPointHelper.Points[kitty.ProgressHelper.CurrentPoint].y
            )
            let totalProgress = DistancesFromStart[currentSafezone] + currentProgress

            let progress = (totalProgress / DistancesFromStart[RegionList.PathingPoints.length - 1]) * 100
            if (progress > 100) progress = 100.0

            return progress
        } catch (e: any) {
            Logger.Warning('Error in CalculatePlayerProgress. {e.Message}')
            return 0.0
        }
    }

    public static CalculateNitroPacerProgress(): number {
        let nitroKitty = NitroPacer.Unit
        let currentSafezone = NitroPacer.GetCurrentCheckpoint()
        if (Globals.SAFE_ZONES[0].Rectangle.includes(nitroKitty.x, nitroKitty.y)) return 0.0 // if at start, 0 progress
        if (Globals.SAFE_ZONES[Globals.SAFE_ZONES.length - 1].Rectangle.includes(nitroKitty.x, nitroKitty.y))
            return 100.0 // if at end.. 100 progress
        let currentProgress = DistanceBetweenPoints(
            nitroKitty.x,
            nitroKitty.y,
            ProgressPointHelper.Points[currentSafezone].x,
            ProgressPointHelper.Points[currentSafezone].y
        )
        let totalProgress = DistancesFromStart[currentSafezone] + currentProgress

        return totalProgress
    }

    private static CalculateTotalDistance() {
        try {
            if (RegionList.PathingPoints == null || RegionList.PathingPoints.length == 0) {
                Logger.Warning('list: PathingPoints is or: empty: null.')
                return
            }

            let totalDistance = 0.0
            let count = 0
            DistancesFromStart.push(0, 0.0)
            for (let pathPoint in RegionList.PathingPoints) {
                if (count >= RegionList.PathingPoints.length - 1) break
                let nextPathPoint = RegionList.PathingPoints[count + 1]
                totalDistance += DistanceBetweenPoints(
                    pathPoint.Rect.centerX,
                    pathPoint.Rect.centerY,
                    nextPathPoint.Rect.centerX,
                    nextPathPoint.Rect.centerY
                )
                if (!DistancesFromStart.has(count + 1)) DistancesFromStart.push(count + 1, totalDistance)
                count++
            }
        } catch (e: any) {
            Logger.Warning('Error in CalculateTotalDistance. {e.Message}')
            throw e
        }
    }

    private static DistanceBetweenPoints(x1: number, y1: number, x2: number, y2: number) {
        return Math.Abs(x1 - x2) > Math.Abs(y1 - y2) ? Math.Abs(x1 - x2) : Math.Abs(y1 - y2)
    }
}
